File: python/ClimbingStairs.py
# 70. Climbing Stairs

#You are climbing a staircase. It takes n steps to reach the top.

#Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?



#Example 1:

#Input: n = 2
#Output: 2
#Explanation: There are two ways to climb to the top.
#1. 1 step + 1 step
#2. 2 steps


class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        memo = {}
        def climb(n):
            if n in memo:
                return memo[n]

            if (n == 1):
                return 1
            if (n == 2):
                return 2
            else:
                result = climb(n - 1) + climb(n - 2)
            memo[n] = result
            print(memo)
            return result

        return climb(n)

File: java/Climbing Stairs.java
/*
 * Climbing Stairs - 70
 */

 class Solution {
  
    // This method calculates the number of distinct ways to climb to the top.
    public int climbStairs(int n) {
        int first = 0, second = 1;
      
        // Loop through number of steps n
        for (int i = 0; i < n; i++) {
            // Calculate next number in the series
            int next = first + second;
          
            // Update the previous two numbers for next iteration
            first = second;
            second = next;
        }
      
        // The 'second' variable holds the total ways to reach the top
        return second;
    }
}